language/evm/hardhat-move/src/compile.ts

```diff
@@ -0,0 +1,253 @@
+    let listRes = await listCompiledContracts(packagePath);
+    if (listRes.isErr()) {
+        return err(new ChainedError(`Failed to list compiled contracts in ${packagePath}`, listRes.error));
+    }
+    let contractNames = listRes.value;
+
+    let genResults = await Promise.all(
+        contractNames.map(contractName => generateArtifact(hardhatRootPath, packagePath, contractName)));
+
+    let errors = [];
+    let artifacts = [];
+    for (let res of genResults) {
+        if (res.isErr()) {
+            errors.push(res.error);
+        }
+        else {
+            artifacts.push(res.value);
+        }
+    }
+
+    if (errors.length > 0) {
+        return err(new ChainedError(`Failed to generate artifacts for ${packagePath}`, errors));
+    }
+
+    return ok(artifacts);
+}
+
+async function buildPackage(arch: string, movePath: string, packagePath: string): Promise<Result<number, MoveBuildError | ChainedError>> {
+    let buildRes = await movePackageBuild(arch, movePath, packagePath);
+    if (buildRes.isErr()) {
+        let e = buildRes.error;
+        console.log(`\nFailed to build ${packagePath}\n${e.stdout}${e.stderr}`);
+        return err(e);
+    }
+
+    console.log(`Successfully built ${packagePath}`);
+    return ok(0);
+}
+
+async function generateArtifacts(hardhatRootPath: string, packagePath: string): Promise<Result<Artifact[], MoveBuildError | ChainedError>> {
+    let genArtifactsRes = await generateArtifactsForPackage(hardhatRootPath, packagePath);
+    if (genArtifactsRes.isErr()) {
+        let e = genArtifactsRes.error;
+        console.log(`Failed to build ${packagePath}\n${e}`);
+        return err(genArtifactsRes.error);
+    }
+
+    console.log(`Successfully generate ABI for ${packagePath}`);
+    return ok(genArtifactsRes.value);
+}
+
+export async function compile(
+    arch: string,
+    movePath: string,
+    artifacts: Artifacts,
+    config: HardhatConfig
+) {
+    let packagePaths: String[] = await listMovePackages(Path.join(config.paths.root, "sources"));
+    packagePaths = [...packagePaths, config.paths.root];
+
+    if (packagePaths.length == 0) {
+        console.log("No Move contracts to compile");
+        return;
+    }
+
+    let plural = packagePaths.length == 1 ? "" : "s";
+    console.log("Building %d Move package%s...", packagePaths.length, plural);
+
+    let buildResults = await Promise.all(
+        packagePaths.map(path => buildPackage(arch, movePath, path.toString())));
+
+    let failedToBuildAll = false;
+    console.assert(packagePaths.length == buildResults.length);
+
+    for (let idx in packagePaths) {
+        let res = buildResults[idx];
+        if (!res.isOk()) {
+            failedToBuildAll = true;
+        }
+    }
+
+    if (failedToBuildAll) {
+        // TODO: terminate gracefully
+        throw new Error("Failed to build one or more Move packages");
+    }
+
+    if (arch != ARCH_ETHEREUM) return;
```

Comment:
If I'm understanding correctly, hardhat is an Ethereum-centered dev environment. Based on this, I'm not sure it makes sense to have arch as a configurable option -- other architectures seem to be out of the scope of this plugin. Could I ask what motivated you to add this?